core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2386,7 +2413,234 @@ protected Scope visitUpdate(Update update, Optional<Scope> scope)
+                        setExpressionTypesBuilder.add(analysis.getType(expression));
+                    }
+                });
+                List<Type> setColumnTypes = setColumnTypesBuilder.build();
+                List<Type> setExpressionTypes = setExpressionTypesBuilder.build();
+                if (!typesMatchForInsert(setColumnTypes, setExpressionTypes)) {
+                    throw semanticException(TYPE_MISMATCH,
+                            operation,
+                            "MERGE table column types don't match for MERGE case %s, SET expressions: Table: [%s], Expressions: [%s]",
+                            caseCounter,
+                            Joiner.on(", ").join(setColumnTypes),
+                            Joiner.on(", ").join(setExpressionTypes));
+                }
+                for (int index = 0; index < caseColumnNames.size(); index++) {
+                    Expression expression = operation.getSetExpressions().get(index);
+                    Type targetType = allColumnTypes.get(caseColumnNames.get(index));
+                    Type expressionType = analysis.getType(expression);
+                    if (!targetType.equals(expressionType)) {
+                        analysis.addCoercion(expression, targetType, typeCoercion.isTypeOnlyCoercion(expressionType, targetType));
+                    }
+                }
+            }
+
+            merge.getMergeCases().stream()
+                    .filter(mergeCase -> mergeCase instanceof MergeInsert)
+                    .findFirst()
+                    .ifPresent(mergeCase -> accessControl.checkCanInsertIntoTable(session.toSecurityContext(), tableName));
+
+            merge.getMergeCases().stream()
+                    .filter(mergeCase -> mergeCase instanceof MergeDelete)
+                    .findFirst()
+                    .ifPresent(mergeCase -> accessControl.checkCanDeleteFromTable(session.toSecurityContext(), tableName));
+
+            Set<String> allUpdateColumnNames = allUpdateColumnNamesBuilder.build();
+            if (!allUpdateColumnNames.isEmpty()) {
+                accessControl.checkCanUpdateTableColumns(session.toSecurityContext(), tableName, allUpdateColumnNames);
+            }
+
+            if (!accessControl.getRowFilters(session.toSecurityContext(), tableName).isEmpty()) {
+                throw semanticException(NOT_SUPPORTED, merge, "Merge table with row filter");
+            }
+
+            for (ColumnSchema column : dataColumnSchemas) {
+                if (!accessControl.getColumnMasks(session.toSecurityContext(), tableName, column.getName(), column.getType()).isEmpty()) {
+                    throw semanticException(NOT_SUPPORTED, merge, "Merge table with column mask");
+                }
+            }
+
+            analysis.setUpdateType("MERGE", tableName, Optional.of(table), Optional.empty());
```

Comment:
Yes, the column list is only used for `UPDATE`.

I've learned a lot since I wrote `UPDATE`, and I'm looking forward to throwing away that code and reimplementing `UPDATE` and `DELETE` as special cases of `MERGE`.

Reply:
I'm not sure of the semantics of this field. However, one of the capabilities of MERGE is to update columns. If the field is never accessed for MERGE, then maybe something is missing?

Reply:
I did this for merge.  I decided there was no harm in returning the union of the columns modified in any merge case.